### PR TITLE
add constants (enumlike) + remove lang form iq

### PR DESCRIPTION
**Pull request by genofire**

### CODE REVIEW

**Comment by genofire** [comment]
which branch?
`builder` or `builder-exploration`

**Comment by mremond** [comment]
`builder` is the branch that should be integrated, with change that are not controversial.
`builder-exploration` contains code that I am not sure about. The change "feels" it could bring structure, but not sure yet if we should do it. Trying to get more feedback :)

If you want, I could merge my `builder` branch so that you can adapt your patch.
Just let me know :)

**Comment by genofire** [comment]
Done
- i force Builder to give a `type` for `iq` and `message`
- remove `lang` on some places ... (the value could be not set)

File: iq.go [hunk 1]
```diff
@@ -137,11 +31,13 @@ type IQPayload interface {
 	Namespace() string
 }
 
-// iqtype, from, to, id, lang string
-func NewIQ(a Attrs) IQ {
+// NewIQ iqtype, attrs
+func NewIQ(t IQType, a Attrs) IQ {
+	// TODO generate IQ ID if not set
```
Comment (genofire) [inline iq.go]: a little todo for you -> if you like.

every iq needs a id, that is importent

**Comment by mremond** [comment]
Thanks, I will have a look :)

File: iq.go (hunk 1 above)
Comment (mremond) [reply iq.go]: Yes, at least ejabberd will not reply without and id.

**Comment by mremond** [comment]
I have some doubts about removing the Type from Attrs.
For example, the type is not mandatory on a message, but in the current state, the type is mandatory on message.

**Comment by genofire** [comment]
yep, so it is already during compiling visible, that you have mistyped

**Comment by mremond** [comment]
I already added the change from presence priority to int in master, as it is obvious.

**Comment by mremond** [comment]
> yep, so it is already during compiling visible, that you have mistyped

The problem is that it make it a bit more cumbersome to set those fields dynamically in the code from a string.
We also would need to solve the question of optional fields. For example, for a type field on a chat message, you can omit it:

> If included, the 'type' attribute MUST have one of the following values: ...

Another approach would be to have a kind if dynamic check of the consistency when you marshal it for example. That was what I was aiming at with the builder: Have a safe way to dynamically check if packet is valid, without being too picky at compile time.

**Comment by mremond** [comment]
> Oh you are right: it could be also something else:
> ` <xs:element name='priority' type='xs:byte'/>`
> https://tools.ietf.org/html/rfc6120#appendix-A.5

I think it cannot be totally anything. More precisely, it s a signed int8 (equivalent to a byte):

> The OPTIONAL <priority/> element contains non-human-readable XML character data that specifies the priority level of the resource. The value MUST be an integer between -128 and +127. 

So, I can move it to int8 instead.

**Comment by mremond** [comment]
Regarding enum, after testing, you can still use a string with anything if you wish to:
https://play.golang.org/p/p0ka7kZZk9H

So, we can still use empty message type. The enum are only helper, but not enforced.
As they are not enforced, maybe we can keep the type in Attrs and use a StanzaType defined as follow:

```go
type StanzaType string

const (
	IQTypeError  StanzaType = "error"
	IQTypeGet    StanzaType = "get"
	IQTypeResult StanzaType = "result"
	IQTypeSet    StanzaType = "set"

	MessageTypeChat      StanzaType = "chat"
	MessageTypeError     StanzaType = "error"
	MessageTypeGroupchat StanzaType = "groupchat"
	MessageTypeHeadline  StanzaType = "headline"
	MessageTypeNormal    StanzaType = "normal"

	PresenceTypeError        StanzaType = "error"
	PresenceTypeProbe        StanzaType = "probe"
	PresenceTypeSubscribe    StanzaType = "subscribe"
	PresenceTypeSubscribed   StanzaType = "subscribed"
	PresenceTypeUnavailable  StanzaType = "unavailable"
	PresenceTypeUnsubscribe  StanzaType = "unsubscribe"
	PresenceTypeUnsubscribed StanzaType = "unsubscribed"
)
```

How does it looks?

**Comment by mremond** [comment]
Another note: An IQ can have a lang set: It can be used by the server to define in which language the form or the errors text will be displayed:
https://xmpp.org/rfcs/rfc6120.html#stanzas-attributes

**Comment by mremond** [comment]
I made a latest commit that gather the enum part under StanzaType.
Hopefully, it is a good balance.

**Comment by mremond** [comment]
Needs update to get the latest changes in master and I can them merge it to master branch.

**Comment by genofire** [comment]
i have it rebased, have fun ;)

**Comment by mremond** [comment]
>i have it rebased, have fun ;)

@genofire many thanks!

**Comment by genofire** [comment]
(was something open? - still not merged ...)

**Comment by mremond** [comment]
>(was something open? - still not merged ...)

@genofire no, all is good. I will merge as soon as possible.

**Comment by mremond** [comment]
Done, thanks!

**Pull request merged**